taprpc/group_key_tapscript_test.go

```diff
@@ -0,0 +1,237 @@
+
+	// Construct the user subtree control block.
+	outputKey := txscript.ComputeTaprootOutputKey(
+		&internalKey, actualTaprootHash1[:],
+	)
+	outputKeyIsOdd := outputKey.SerializeCompressed()[0] == 0x03
+
+	inclusionProof := bytes.Join(
+		[][]byte{
+			assetIDLeafHash[:],
+		}, nil,
+	)
+
+	userSubtreeControlBlock := txscript.ControlBlock{
+		InternalKey:     &internalKey,
+		OutputKeyYIsOdd: outputKeyIsOdd,
+		LeafVersion:     txscript.BaseLeafVersion,
+		InclusionProof:  inclusionProof,
+	}
+
+	// Ensure the custom script control block is correct by computing the
+	// root hash given the control block and the custom script leaf.
+	rootCheckBytes := userSubtreeControlBlock.RootHash(
+		customScriptLeaf.Script,
+	)
+
+	var rootCheck chainhash.Hash
+	copy(rootCheck[:], rootCheckBytes)
+
+	require.Equal(t, actualTaprootHash1, rootCheck)
+
+	// ---------------------------------------------------------------------
+	//
+	// Example with user-provided branch only.
+
+	// Construct a branch from two node hashes. (A node hash can be a leaf
+	// hash or a branch hash.) We must use two nodes so that we can be sure
+	// that there is only one asset ID leaf in layer 1.
+	//
+	// In this case, we use the customScriptLeaf as one of the nodes so that
+	// we can test the control block.
+	userLeafHash := customScriptLeaf.TapHash()
+	userSiblingNodeHash := chainhash.Hash([32]byte{3, 4, 5})
+
+	userBranch := &TapBranch{
+		LeftTaphash:  userLeafHash[:],
+		RightTaphash: userSiblingNodeHash[:],
+	}
+	branchNodes, err := UnmarshalTapscriptBranch(userBranch)
+	require.NoError(t, err)
+
+	branchPreimage, err := commitment.NewPreimageFromTapscriptTreeNodes(
+		*branchNodes,
+	)
+	require.NoError(t, err)
+
+	userTreeRootHash, err = branchPreimage.TapHash()
+	require.NoError(t, err)
+
+	actualTaprootHash2 := tapBranchHash(assetIDLeafHash, *userTreeRootHash)
+	fmt.Printf("Taproot hash with user branch: %v\n", actualTaprootHash2)
+
+	// Construct the user subtree control block. This block targets the
+	// custom script leaf.
+	outputKey = txscript.ComputeTaprootOutputKey(
+		&internalKey, actualTaprootHash2[:],
+	)
+	outputKeyIsOdd = outputKey.SerializeCompressed()[0] == 0x03
+
+	inclusionProof = bytes.Join(
+		[][]byte{
+			userSiblingNodeHash[:],
+			assetIDLeafHash[:],
+		}, nil,
+	)
+
+	userSubtreeControlBlock = txscript.ControlBlock{
+		InternalKey:     &internalKey,
+		OutputKeyYIsOdd: outputKeyIsOdd,
+		LeafVersion:     txscript.BaseLeafVersion,
+		InclusionProof:  inclusionProof,
+	}
+
+	// Ensure the custom script control block is correct by computing the
+	// root hash given the control block and the custom script leaf.
+	rootCheckBytes = userSubtreeControlBlock.RootHash(
+		customScriptLeaf.Script,
+	)
+	copy(rootCheck[:], rootCheckBytes)
+
+	require.Equal(t, actualTaprootHash2, rootCheck)
+
+	// ---------------------------------------------------------------------
+	//
+	// Example with an optional user-provided root hash only.
```

Comment:
Nice, thank you for outlining the three different approaches.

IMO the decision boils down to:
Proof size, privacy and BIP complexity: The "preimage based" approach is slightly larger and would expose the script (if a single one is used) in the reveal, but describing it would be slightly easier in the BIP (since we already have the tapscript sibling approach in other places).

So I'm okay with either of them. Thanks for your patience and persistence in explaining your approach to me :see_no_evil: